From: Benjamin Drung <benjamin.drung@canonical.com>
Date: Thu, 19 Mar 2026 16:11:37 +0100
Subject: fix(base): move initrd.target.wants symlink creation to wait_for_dev

Commit 4d9d767da2e5 ("crypt/parse-crypt.sh: hide encrypted devices from
systemd timeout warnings.") split parts of `wait_for_dev` into
`set_systemd_timeout_for_dev` to allow setting the systemd timeout
separately. `parse-crypt.sh` is the only direct caller of
`set_systemd_timeout_for_dev`.

Fixes: 4d9d767da2e5 ("crypt/parse-crypt.sh: hide encrypted devices from systemd timeout warnings.")
Forwarded: https://github.com/dracut-ng/dracut-ng/pull/2310
---
 modules.d/80base/dracut-dev-lib.sh | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/modules.d/80base/dracut-dev-lib.sh b/modules.d/80base/dracut-dev-lib.sh
index 63c2c19..7b71801 100755
--- a/modules.d/80base/dracut-dev-lib.sh
+++ b/modules.d/80base/dracut-dev-lib.sh
@@ -75,12 +75,6 @@ set_systemd_timeout_for_dev() {
     _timeout=${_timeout:-infinity}
 
     _name=$(dev_unit_name "$1")
-    if ! [ -L "${PREFIX-}/etc/systemd/system/initrd.target.wants/${_name}.device" ]; then
-        [ -d "${PREFIX-}"/etc/systemd/system/initrd.target.wants ] || mkdir -p "${PREFIX-}"/etc/systemd/system/initrd.target.wants
-        ln -s ../"${_name}".device "${PREFIX-}/etc/systemd/system/initrd.target.wants/${_name}.device"
-        type mark_hostonly > /dev/null 2>&1 && mark_hostonly /etc/systemd/system/initrd.target.wants/"${_name}".device
-        _needreload=1
-    fi
 
     if ! [ -f "${PREFIX-}/etc/systemd/system/${_name}.device.d/timeout.conf" ]; then
         mkdir -p "${PREFIX-}/etc/systemd/system/${_name}.device.d"
@@ -125,6 +119,15 @@ wait_for_dev() {
         printf 'warn "\"%s\" does not exist"\n' "$1"
     } >> "${PREFIX-}$hookdir/emergency/80-${_name}.sh"
 
+    [ -n "${DRACUT_SYSTEMD-}" ] || return 0
+
+    if ! [ -L "${PREFIX-}/etc/systemd/system/initrd.target.wants/${_name}.device" ]; then
+        [ -d "${PREFIX-}"/etc/systemd/system/initrd.target.wants ] || mkdir -p "${PREFIX-}"/etc/systemd/system/initrd.target.wants
+        ln -s ../"${_name}".device "${PREFIX-}/etc/systemd/system/initrd.target.wants/${_name}.device"
+        type mark_hostonly > /dev/null 2>&1 && mark_hostonly /etc/systemd/system/initrd.target.wants/"${_name}".device
+        /sbin/initqueue --onetime --unique --name daemon-reload systemctl daemon-reload
+    fi
+
     set_systemd_timeout_for_dev $_noreload "$@"
 }
 
